### ISSUE TITLE
DataEditR: An Interactive R Package for Viewing, Entering & Editing Data

### ISSUE THREAD

**Issue by DillonHammill** [ropensci/software-review#391]
Submitting Author: @DillonHammill
Repository:   https://github.com/DillonHammill/DataEditR
Version submitted:   0.0.5
Editor: TBD  
Reviewer 1: TBD  
Reviewer 2: TBD  
Archive: TBD  
Version accepted: TBD   

---

-   Paste the full DESCRIPTION file inside a code block below:

```
Package: DataEditR
Title: Interactive Data Viewing, Entry and Editing
Version: 0.0.1
Date: 2020-07-21
Authors@R: 
    person(given = "Dillon",
           family = "Hammill",
           role = c("aut", "cre"),
           email = "Dillon.Hammill@anu.edu.au",
           comment = c(ORCID = "0000-0002-1407-7223"))
Description: Provides an interactive interface to view, enter or edit data 
  or tabular data files in R.
URL: https://github.com/DillonHammill/DataEditR
BugReports: https://github.com/DillonHammill/DataEditR/issues
Depends:
  R(>= 3.5.0)
Imports:
  shiny,
  shinythemes,
  rhandsontable,
  htmltools
License: GPL-2
Encoding: UTF-8
LazyData: true
Roxygen: list(markdown = TRUE)
RoxygenNote: 7.1.1
Language: en-GB
Suggests: 
    knitr,
    rmarkdown
VignetteBuilder: knitr
```

## Scope 

- Please indicate which category or categories from our [package fit policies](https://ropensci.github.io/dev_guide/policies.html#package-categories) this package falls under: (Please check an appropriate box below. If you are unsure, we suggest you make a pre-submission inquiry.):

	- [ ] data retrieval
	- [ ] data extraction
	- [x] data munging
	- [ ] data deposition
	- [ ] workflow automation
	- [ ] version control
	- [ ] citation management and bibliometrics
	- [ ] scientific software wrappers
	- [ ] field and lab reproducibility tools
	- [ ] database software bindings
	- [ ] geospatial data
	- [ ] text analysis
	
- Explain how and why the package falls under these categories (briefly, 1-2 sentences):

`DataEditR` provides an interactive interface built on `rhandsontable` that supports viewing, entering and editing of data in R without the need for any coding. `DataEditR` allows users to interactively construct data.frames from scratch and edit/view existing data objects (e.g. matrices, data.frames and data.tables) or data saved to file (e.g. csv - any tabular data format is supported).

-   Who is the target audience and what are scientific applications of this package?  

`DataEditR` has been designed to be as simple as possible to appeal to users with limited coding experience. It provides a familiar interface for viewing, entering and editing data that users of GUI-oriented software will be familiar with. `DataEditR` is also extremely useful for developers seeking an interactive interface to obtain inputs from the user. This package therefore has broad applications for any discipline that requires manual data entry and editing.

-   Are there other R packages that accomplish the same thing? If so, how does yours differ or meet [our criteria for best-in-category](https://ropensci.github.io/dev_guide/policies.html#overlap)?

The `edit()` function from the `utils` package also offers the ability to interactively enter/edit data but lacks a lot of the features that one would expect from a GUI-oriented data manipulation tool (e.g. drag to fill, addition of checkboxes and dropdown menus). Not to mention the UI is not very visually appealing and there are undesirable glitches when entering data.

Perhaps the most similar packages to `DataEditR` that offer similar features are `DT` and `rhandsontable` (on which `DataEditR` is built). To the best of my knowledge, `DT` lacks the ability to interactively add/remove columns and rows and does not support editing of row/column names. For the most part, these packages were designed to interactively visualize datasets, whilst the focus the `DataEditR` is data entry and editing. `rhandsontable` has a lot of support for more complex visualisations of data, but the inner workings are quite complex (transferring data between R and javascript means that there are specific requirements for certain features to work). Currently, `rhandsontable` does not support editing of column names, this has been implemented in `DataEditR` through addition of custom javascript code. `DataEditR` has been developed over time to retain as much of the features of `rhandsontable` as possible whilst taking care of all the required data formatting internally. 

The integration of reading/writing capabilities within `DataEditR` means that all the data loading, editing and saving operations can be performed in a single line (to minimize coding). To the best of my knowledge, `DataEditR` is the only R package that lets users interactively create data.frames from scratch without any coding.

-   (If applicable) Does your package comply with our [guidance around _Ethics, Data Privacy and Human Subjects Research_](https://devguide.ropensci.org/policies.html#ethics-data-privacy-and-human-subjects-research)?

-   If you made a pre-submission enquiry, please paste the link to the corresponding issue, forum post, or other discussion, or @tag the editor you contacted.

## Technical checks

Confirm each of the following by checking the box.  

- [x] I have read the [guide for authors](https://devguide.ropensci.org/guide-for-authors.html) and [rOpenSci packaging guide](https://devguide.ropensci.org/building.html).

This package:

- [x] does not violate the Terms of Service of any service it interacts with. 
- [x] has a CRAN and OSI accepted license.
- [x] contains a [README with instructions for installing the development version](https://ropensci.github.io/dev_guide/building.html#readme). 
- [x] includes [documentation with examples for all functions, created with roxygen2](https://ropensci.github.io/dev_guide/building.html#documentation).
- [x] contains a vignette with examples of its essential functions and uses.
- [ ] has a [test suite](https://ropensci.github.io/dev_guide/building.html#testing).
- [x] has [continuous integration](https://ropensci.github.io/dev_guide/ci.html), including reporting of test coverage using services such as Travis CI, Coveralls and/or CodeCov.

## Publication options

- [x] Do you intend for this package to go on CRAN?  
- [ ] Do you intend for this package to go on Bioconductor?  
- [ ] Do you wish to automatically submit to the [Journal of Open Source Software](http://joss.theoj.org/)? If so:

<details>
 <summary>JOSS Options</summary>  

  - [ ] The package has an **obvious research application** according to [JOSS's definition](https://joss.readthedocs.io/en/latest/submitting.html#submission-requirements).
    - [ ] The package contains a `paper.md` matching [JOSS's requirements](https://joss.readthedocs.io/en/latest/submitting.html#what-should-my-paper-contain) with a high-level description in the package root or in `inst/`.
    - [ ] The package is deposited in a long-term repository with the DOI: 
    - (*Do not submit your package separately to JOSS*)  
  
</details>

- [ ] Do you wish to submit an Applications Article about your package to [Methods in Ecology and Evolution](http://besjournals.onlinelibrary.wiley.com/hub/journal/10.1111/(ISSN)2041-210X/)? If so:  

<details>
<summary>MEE Options</summary>  

- [ ] The package is novel and will be of interest to the broad readership of the journal. 
- [ ] The manuscript describing the package is no longer than 3000 words.
- [ ] You intend to archive the code for the package in a long-term repository which meets the requirements of the journal (see [MEE's Policy on Publishing Code](http://besjournals.onlinelibrary.wiley.com/hub/journal/10.1111/(ISSN)2041-210X/journal-resources/policy-on-publishing-code.html))
- (*Scope: Do consider MEE's [Aims and Scope](http://besjournals.onlinelibrary.wiley.com/hub/journal/10.1111/(ISSN)2041-210X/aims-and-scope/read-full-aims-and-scope.html) for your manuscript. We make no guarantee that your manuscript will be within MEE scope.*)
- (*Although not required, we strongly recommend having a full manuscript prepared when you submit here.*)
- (*Please do not submit your package separately to Methods in Ecology and Evolution*)

</details>

## Code of conduct

- [x] I agree to abide by [rOpenSci's Code of Conduct](https://ropensci.github.io/dev_guide/policies.html#code-of-conduct) during the review process and in maintaining my package should it be accepted.

**Comment by geanders** [comment]
Thanks so much for your submission, @DillonHammill! It looks like a very interesting package, and one that will be useful for many R users.

I have two follow up questions: 

First, a question regarding if the package is in-scope for rOpenSci: The package certainly looks like it could fall within "data munging", which is one of our categories. However, we also now have a special consideration for packages that are interactive: 

> "Packages that include interactive tools to facilitate researcher workflows (e.g., shiny apps) must have a mechanism to make the interactive workflow reproducible, such as code generation or a scriptable API." (from Section 5.2.2 of the [rOpenSci Packages book](https://devguide.ropensci.org/policies.html#aims-and-scope))

For the purposes of this package, I think this would mean generating a code script that replicates all the interactive actions that the user has taken in data munging through the interactive interface the package provides. From the package README file, I didn't see any indication that a script is generated that could be used to replicate the interactive actions taken by the user. Could you please let me know if I've missed anything there?

Second, we typically ask for at least 75% coverage of package functions by unit tests, or an explanation for why coverage is lower (see Section 1.9 of the [rOpenSci Packages book](https://devguide.ropensci.org/building.html#testing)). It looks like this package does not have a testing suite. If the package is in-scope, we would need you to add a justification for the lack of a testing suite before the review.

Thanks very much for any details you can provide on either of these points!

**Comment by DillonHammill** [comment]
Thanks for taking a look at DataEditR @geanders. I hadn't thought about generating code to mimic the interactive edits, but it is certainly possible.

I haven't included unit tests at the moment as it is only a single function that is widely used in CytoExploreR. This is where I would pick up on any problems. I am working on adding some shiny tests but from memory these are not included in the test coverage anyway? I will also need to re-structure the package in order to be able to write shinytests for this package. 

I will work on these and get back to you ASAP.

**Comment by melvidoni** [comment]
Hello @DillonHammill. The EIC rotated, and I'm taking over this submission. Were you able to address this questions?

**Comment by DillonHammill** [comment]
Hi @melvidoni, I have had a few requests to convert the code to use shiny modules (https://github.com/DillonHammill/DataEditR/issues/5) so that the package can be used within other shiny applications and so I am currently working on this.

I have added a testing suite to the package but this will need to be updated once I finish making the shiny module changes.

In terms of returning the code to mimic the interactive edits made to the data, I think it is possible and I am exploring a number of options:

- Simply return the code required to generate the final edited version of the data. This is the simplest approach but does not display the steps taken to arrive at the final edited version.
- Record all data edits to a temp file that can then be converted into code once the application is closed. I don't think that this is the way to go, as users tend to get carried away when things are interactive and many of the changes made need not be reflected in the code output. There are some complex challenges to get this to work and this is why rhandsontable uses the approach above to output the code.
- Another alternative is that I simply return the code required to convert the starting data to the returned data. This is much simpler as I would only need to compare the initial and final `data.frames` and convert changes to code. 

I am exploring ways in which I can use a combination of options 2 & 3 to generate the code in an efficient manner. It will take some time for me to properly implement this, so I am happy to resubmit or put this on hold until I have figured this out.
